internal/expr/bindinputs.go

```diff
@@ -126,13 +65,29 @@ func newOutputColumn(tableName string, columnName string, output typeinfo.Output
 	return outputColumn{column: tableName + "." + columnName, output: output}
 }
 
+type typedExpr interface {
+	addToQueryBuilder(*queryBuilder) error
+}
+
 // typedInputExpr stores information about a Go value to use as a standalone query
 // input.
 type typedInputExpr struct {
 	input typeinfo.Input
 }
 
-// insertColumn stores information about a single column of a row in an INSERT
+// addToQueryBuilder adds the typed input expressions to the query builder.
+func (te *typedInputExpr) addToQueryBuilder(qb *queryBuilder) error {
+	return qb.addInput(te)
```

Comment:
This is not really _visitor_, because `queryBuilder` holds all the logic, and if you add another `typedExpr` implementation that it needs to handle, it grows.

That said, this _is_ managing complexity after a fashion, though you've tightly coupled the builder and the expressions.

Reply:
One way to do it would be for `queryBuilder` to implement an interface that exposes what the visitors need, then push the logic back into the visitors. Looking at:
```go
func (qb *queryBuilder) addInput(te *typedInputExpr) error {
	params, err := te.input.LocateParams(qb.typeToValue)
	if err != nil {
		return err
	}
	if params.Omit {
		return omitEmptyInputError(te.input.Desc())
	}
	if params.Bulk {
		return fmt.Errorf("cannot use bulk inputs outside an insert statement")
	}

	qb.argUsed[params.ArgTypeUsed] = true
	qb.sqlBuilder.writeInput(qb.inputCount, len(params.Vals))
	for _, val := range params.Vals {
		namedInput := sql.Named("sqlair_"+strconv.Itoa(qb.inputCount), val)
		qb.namedInputs = append(qb.namedInputs, namedInput)
		qb.inputCount++
	}
	return nil
}
```
The first half is validation that could (and should) be on `typedInputExpr`.
The rest could be done by coordinating methods like `setArgTypeUsed`, `writeInput`, `addNamedInput` or some such.